eth/transaction_signing.go

```diff
@@ -0,0 +1,121 @@
+package eth
+
+import (
+	"encoding/hex"
+	"errors"
+
+	secp256k1 "github.com/btcsuite/btcd/btcec"
+
+	"github.com/INFURA/go-ethlibs/rlp"
+)
+
+var ErrInsufficientParams = errors.New("transaction is missing values")
+var ErrFatalCrytpo = errors.New("unable to sign Tx with private key")
+
+// Looks like there have been multiple attempts to get the Koblitz curve (secp256k1) supported in golang
+//
+// https://github.com/golang/go/pull/26873 <-- rejected
+// https://github.com/golang/go/issues/26776 <-- rejected
+// using "github.com/btcsuite/btcd/btcec"
+// other alternative is to import the C library but want to avoid that if possible
+
+func (t *Transaction) Sign(privateKey string, chainId uint64) (string, error) {
+
+	if !t.check() {
+		return "", ErrInsufficientParams
+	}
+
+	pKey, err := hex.DecodeString(privateKey)
+	if err != nil {
+		return "", err
+	}
+
+	rawTx, err := t.serialize(chainId, false)
+	if err != nil {
+		return "", err
+	}
+
+	privKey, pubKey := secp256k1.PrivKeyFromBytes(secp256k1.S256(), pKey)
+	signature, err := privKey.Sign(rawTx)
+	if err != nil {
+		return "", err
+	}
+
+	verified := signature.Verify(rawTx, pubKey)
+	if !verified {
+		panic(ErrFatalCrytpo)
+	}
+
+	// extract signature (r,s,v)
+	t.signatureValues(signature)
+	signed, err := t.serialize(chainId, true)
+	return string(signed), err
+}
+
+func (t *Transaction) serialize(chainId uint64, signature bool) ([]byte, error) {
+
+	// my brain has turned to mush, pretty sure this not the most efficient way
+	// but trying to follow ether.js logic
+	var list []rlp.Value
+	list = append(list, t.Nonce.RLP())
+	list = append(list, t.GasPrice.RLP())
+	list = append(list, t.Gas.RLP())
+	list = append(list, t.To.RLP())
+	list = append(list, t.Value.RLP())
+	list = append(list, rlp.Value{String: t.Input.String()})
+
+	var temp []byte
+	if !signature {
+		if chainId != 0 {
+			list = append(list, QuantityFromUInt64(chainId).RLP())
+			r, err := rlp.From("0x")
+			if err != nil {
+				return temp, err
+			}
+			list = append(list, *r)
+			s, err := rlp.From("0x")
+			if err != nil {
+				return temp, err
+			}
+			list = append(list, *s)
+		}
+		rawTx := rlp.Value{List: list}
+		h, err := rawTx.HashToBytes()
+		return h, err
+	} else {
+		if chainId != 0 {
+			v := chainId*2 + 8
+			newV := t.V.UInt64() + v
+			t.V = QuantityFromUInt64(newV)
+		}
+		list = append(list, t.V.RLP())
+		list = append(list, t.R.RLP())
+		list = append(list, t.S.RLP())
+		rawTx := rlp.Value{List: list}
+
+		signed, err := rawTx.Encode()
+		return []byte(signed), err
```

Comment:
I'm pretty confused what this function's output is supposed to be.  Here it's returning the RLP hexadecimal string data cast to a byte slice, while the other case above seems to be returning the tx hash as a byte slice.

I think it'd be much cleaner if the function always returned the data in the same format rather than dependent on the `signature` input bool (which should renamed to be more clear what it's controlling over).

Reply:
Ideally this function could be refactored out into a stand-alone `ECSign` function to parallel the `ECRecover` used in tx decoding.

The signature for this function could be:

```golang
type Signature struct {
  V, R, S Quantity
}

func ECSign(hash *Hash, privateKey *Data, chainId Quantity) (*Signature, error) { ... } 
```

I feel like doing that would make the logic easier to follow, and parallel other projects like: https://github.com/ethereumjs/ethereumjs-util/blob/master/src/signature.ts#L15

Reply:
I change the result to rlp.Value, I still have stuff commented out but hopefully it's clearer what values are being added depending on whether or not signature is enabled or not

Reply:
IMHO, I think that depends on which library you have been looking at 😓. I've been looking at : https://github.com/ethers-io/ethers.js/blob/c2ce59f95e2b53cb4121c9e941a312913f4fc279/utils/transaction.js#L44 which makes more sense to me than the ethereumjs stuff.. I have been looking more than ethereumjs now but after looking at 3 different libraries that have done transaction signing, I don't know there's a clear winner on how best to do it. I could change all the code around but again, i don't know if it's making any clearer or just moving stuff around into different functions and different return values...   🤷

Reply:
Ok, makes much more sense returning an RLP value!  With this change and my ECSign function in #24, I think the rest of `Transaction.Sign(...)` becomes:

```golang
 	rawTx, err := serialize(chainId, false)
 	if err != nil { ... }
 	signature, err := ECSign(rawTx.Hash(), pKey, chainId)
 	if err != nil { ... } 
 	t.R = signature.R
        t.S = signature.S
        t.V = signature.V
        return t.Encode() 
```

Reply:
Oh! And just realized that the `signature = true` path wouldn't be needed any more, since that's basically doing the same thing as `Transaction.Encode()` at that point, so `serialize` can be simplified to just the `false` case.

Reply:
serialize creates a new rlp.Value by extracting values out of Tx. I still need to add the R,S,V values to the rlp.Value so I can't get rid of signature = true using the current method. But if you really don't like the signature = true stuff, I can return an array of rlp.Values as the return values to serialize, manually add the R,S,V values to the list and create a new rlp.Value out of the list but again, feels like i'm just moving code around

Reply:
Sorry, I thought we already had a function that returned the Transaction RLP, but we don't.  So I think the best solution is:

- Add a `Transaction.RLP()` that returns an `RLP{List:}` of all the values in consensus-order.  It's up to you if it uses `serialize` to do that, but the one thing weird is that when `signature` is `false` then the `chainId` isn't used, so the function feels a bit clunky.

Adjust `.Sign()` like so:

```golang
 	rawTx, err := t.serialize(chainId, false)
 	if err != nil { ... }
 	signature, err := ECSign(rawTx.Hash(), pKey, chainId)
 	if err != nil { ... } 
 	t.R = signature.R
        t.S = signature.S
        t.V = signature.V
        return t.RLP().Encode()
```

This has the advantage that `Transaction.RLP()` can be reused in other places beyond signing.